Replace all uses of deprecated symbol_exprt constructors [blocks: #3768]

tautschnig opened the pull request:
This helps type safety has it avoids constructing symbol_exprts that never get
a proper type (or identifier).

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 115efb34e0442462fd5cef19dc2cd0be2f4596e1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97091511

kroening commented on jbmc/src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -170,7 +170,8 @@ void java_simple_method_stubst::create_method_stub(symbolt &symbol)
-      init_symbol_expression, symbol_exprt(this_argument.get_identifier()));
+      init_symbol_expression,
+      symbol_exprt(this_argument.get_identifier(), this_argument.type()));
     get_argument.add_source_location() = synthesized_source_location;
```
This one does appear to be a bugfix!

tautschnig replied:
Yes, but it should actually get merged via #3747, I'll then rebase this PR on top of it.

kroening commented on jbmc/unit/solvers/refinement/string_constraint_instantiation/instantiate_not_contains.cpp:
```diff
@@ -215,7 +215,7 @@ SCENARIO("instantiate_not_contains",
         simplify(sc.s1, empty_ns);
-        witnesses[sc] = generator.fresh_symbol("w", t.witness_type());
+        witnesses.emplace(sc, generator.fresh_symbol("w", t.witness_type()));
         nc_axioms.push_back(sc);
```
Not sure that the .emplace adds a lot here (exprt is very cheap to construct).

tautschnig replied:
The `.emplace` is necessary, because a `symbol_exprt` cannot be default constructed.

kroening commented on src/analyses/locals.cpp:
```diff
@@ -21,12 +21,13 @@ void localst::build(const goto_functiont &goto_function)
+    locals_map.emplace(
+      param.get_identifier(),
+      symbol_exprt(param.get_identifier(), param.type()));
 }
```
Same here

kroening commented on src/goto-instrument/concurrency.cpp:
```diff
@@ -58,14 +58,14 @@ class concurrency_instrumentationt
     typet type;
-    symbol_exprt array_symbol, w_index_symbol;
+    exprt array_symbol, w_index_symbol;
   };
```
Can these ever be something that's not a symbol_exprt?
If initialisation is delayed, then perhaps optionalt<symbol_exprt> is the right way to say that.

tautschnig replied:
Well, the code is not particularly complete. These are actually never set at the moment.

kroening commented on src/goto-symex/symex_function_call.cpp:
```diff
@@ -409,7 +409,7 @@ void goto_symext::locality(
     {
-      frame.old_level1[l0_name]=c_it->second;
+      frame.old_level1.emplace(l0_name, c_it->second);
       c_it->second = std::make_pair(ssa, frame_nr);
```
Same here

kroening commented on src/util/ssa_expr.h:
```diff
@@ -16,15 +16,9 @@ Author: Daniel Kroening, kroening@kroening.com
-  explicit ssa_exprt(const exprt &expr):
-    symbol_exprt(expr.type())
+  explicit ssa_exprt(const exprt &expr) : symbol_exprt(ID_nil, expr.type())
   {
```
This is the first time you've used the nil ID -- before it has been the empty one.
If you are concerned about any time taken to convert "" into an irep_idt, then use irep_idt().

tautschnig replied:
Now using `irep_idt()` here.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2a5c14aa579799c03bb2b55362c6c1717b5994e5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97291629

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b71d8f91adbb941982c5970135d0ea5dec581bee).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97316228

romainbrenguier commented on jbmc/src/java_bytecode/java_utils.cpp:
```diff
@@ -285,7 +285,8 @@ exprt make_function_application(
   // Function application
-  return function_application_exprt(symbol_exprt(fun_name), arguments, type);
+  return function_application_exprt(
+    symbol_exprt(fun_name, typet()), arguments, type);
```
This function doesn't seem correct. I think it should be changed to take a `code_typet` as argument representing the type of the function instead of `type` (the return type can be deduced from the `code_typet`

tautschnig replied:
So I would have thought, but this does not seem to be the case for a `function_application_exprt`, or else what the Java front-end is doing is completely broken. @kroening maybe?

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_convert_method_class.h:
```diff
@@ -119,7 +119,8 @@ class java_bytecode_convert_methodt:public messaget
 
-    variablet() : symbol_expr(), start_pc(0), length(0), is_parameter(false)
+    variablet()
+      : symbol_expr("", typet()), start_pc(0), length(0), is_parameter(false)
```
Would it make sense for the type to be an argument of the constructor?

romainbrenguier commented on src/goto-symex/slice_by_trace.cpp:
```diff
@@ -31,10 +31,6 @@ void symex_slice_by_tracet::slice_by_trace(
-  merge_identifier="goto_symex::\\merge";
-  merge_symbol=symbol_exprt(typet(ID_bool));
-  merge_symbol.set_identifier(merge_identifier);
-
```
This should have been in a separate commit, so that you can explain this specific change

romainbrenguier commented on src/goto-symex/slice_by_trace.h:
```diff
@@ -17,9 +17,11 @@ Author: Alex Groce (agroce@gmail.com)
+    : ns(_ns),
+      alphabet_parity(false),
+      merge_symbol("goto_symex::\\merge", typet(ID_bool))
+
```
extra blank line

romainbrenguier commented on src/solvers/smt2/smt2_parser.h:
```diff
@@ -46,6 +46,11 @@ class smt2_parsert
+      : term(_term), name(_name)
+    {
+    }
+
```
should have been in a separate commit to explain why this change is needed

martin-cs replied:
+1 this is not clear.

tautschnig replied:
I have added a comment (default-constructing a `symbol_exprt` is deprecated, thus the `name` member needs to be initialised upon construction). Goes together with the use of `emplace` in smt2_parser.cpp.

romainbrenguier reviewed (commented):
Would have been better if split into several commits. `typet()` is used in several places so I'm not sure how this changes are useful.

smowton commented:
I get that you want to avoid people casually saying `symbol_exprt("hello")` and stumble into a dangerous part-constructed expression without noticing -- but as Romain notes, this is just being replaced with explicit `typet()`. That being so, how about moving the evil constructor behind a static function? For example:

```
private:
  // Partial constructor, to be avoided where possible. Use via `::typeless`
  symbol_exprt(const irep_idt &id) { ... }

public:
  // Prefer the public constructor whenever possible.
  static symbol_exprt typeless(const irep_idt &id) { return symbol_exprt(id); }
```

tautschnig commented:
I'll try to implement @smowton's suggestion as soon as possible.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ab2ae0ebfede21b2cbede41f101fb79e269c855c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98019728

tautschnig committed:
Replace all uses of deprecated symbol_exprt constructors

This helps type safety has it avoids constructing symbol_exprts that never get a
proper type (or identifier).

tautschnig commented:
I have started to refactor the commit into smaller ones and have created first PRs that just contain those.

tautschnig commented:
Separate PRs created where the individual bits should now cover all the feedback received.

tautschnig commented:
This PR now only has a single commit left, (re-)reviews would be appreciated.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c539028f83d56857b54e5838958b189d952d5cc3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98341703

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 45c74ceb269b65e34441e6e3785ad993a31e2c3d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98712690

martin-cs reviewed (approved):
The bits that touch my areas of responsibility seem fine.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: d16fbebca858fd37275d17025d5a88567c421c1b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98896786
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 267094c7eb4bd9cca8b3bcae22e60a9b67d86a7f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99079556
